presto-hive/src/main/java/com/facebook/presto/hive/util/AsyncQueue.java

```diff
@@ -43,14 +56,33 @@
      *
      * @throws NullPointerException if the element is null
      */
-    public void add(T element)
+    public Optional<CompletableFuture<?>> add(T element)
     {
         requireNonNull(element, "element is null");
-        if (finished.get()) {
-            return;
-        }
+        checkState(!finished.get());
```

Comment:
I think there's actually a legitimate case where `add()` might be called after `finish()`. `HiveSplitSource` has a `close()` method, which calls `queue.finish()` and then `loader.stop()`, this means that briefly the loader tasks will still be running while the queue is "closed". Either `HiveSplitLoader.stop()` needs to be made synchronous so that it wait for all the tasks to terminate, or this needs to allow for the case when `add()` is called after `finish()` in which case it seems like it should ignore the element and return `Optional.empty()`